pennylane/templates/state_preparations/basis_qutrit.py

```diff
@@ -0,0 +1,116 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+Contains the BasisStatePreparation template.
+"""
+
+import pennylane as qml
+from pennylane.operation import Operation, AnyWires
+
+
+class QutritBasisStatePreparation(Operation):
+    r"""
+    Prepares a basis state on the given wires using a sequence of TShift gates.
+
+    .. warning::
+
+        ``basis_state`` influences the circuit architecture and is therefore incompatible with
+        gradient computations.
+
+    Args:
+        basis_state (array): Input array of shape ``(n,)``, where n is the number of wires
+            the state preparation acts on.
+        wires (Iterable): wires that the template acts on
+
+    **Example**
+
+    .. code-block:: python
+
+        dev = qml.device("default.qutrit", wires=4)
+
+        @qml.qnode(dev)
+        def circuit(basis_state, obs):
+            qml.QutritBasisStatePreparation(basis_state, wires=range(4))
+            return [qml.expval(qml.THermitian(wires=i)) for i in range(4)]
+
+        basis_state = [0, 1, 1, 0]
+        obs = np.array([[1, 1, 0], [1, -1, 0], [0, 0, np.sqrt(2)]]) / np.sqrt(2)
+        print(circuit(basis_state, obs)) # [array(0.70710678), array(-0.70710678), array(-0.70710678), array(0.70710678)]
+    """
+    num_params = 1
+    num_wires = AnyWires
+    grad_method = None
+
+    def __init__(self, basis_state, wires, do_queue=True, id=None):
+        basis_state = qml.math.stack(basis_state)
+
+        # check if the `basis_state` param is batched
+        batched = len(qml.math.shape(basis_state)) > 1
+
+        state_batch = basis_state if batched else [basis_state]
+
+        for i, state in enumerate(state_batch):
+            shape = qml.math.shape(state)
+
+            if len(shape) != 1:
+                raise ValueError(
+                    f"Basis states must be one-dimensional; state {i} has shape {shape}."
+                )
+
+            n_bits = shape[0]
+            if n_bits != len(wires):
+                raise ValueError(
+                    f"Basis states must be of length {len(wires)}; state {i} has length {n_bits}."
+                )
+
+            if not qml.math.is_abstract(state):
```

Comment:
Oh, thank you for the explanation about `is_abstract`! I was actually not sure what the function was doing and I could not access the docs of the qml.math module for some reason. For future references, do you know how to get access to it? Also, if JIT is not compatible with `copmute_decomposition` in general, should we also erase this check from the qubit version: `BasisStatePreparation`?

Reply:
No worries! Basically, when jitting, you try to make an abstract version of the full function, with placeholders for all the variables that won't be set until the function is called. Then you execute all of the logic once (with the placeholders). After that, you can pass it variables to put instead of the placeholders to get the results. This makes the first execution slow, because it needs to compile everything in a generalized fashion, but makes all subsequent calls much faster, because _almost_ everything is already done. 

So `is_abstract` if a function we use to check if we are currently jitting - we just take one of the objects that we expect to be a represented by a placeholder when jitting, and check if it is a concrete value or a placeholder (abstract). For future reference, you can find the math module documentation under API on the Pennylane.ai documentation page! https://docs.pennylane.ai/en/stable/code/qml_math.html

Wrt whether this is an issue for the qubit implementation, the jitting problem in this case is specific to _this_ implementation of `compute_decomposition`. The qubit version is JIT compatible, because it has a separate clause for generating a tape with parametrized operators if `is_abstract` returns `True`. Basically, operator _parameters_ can be abstract, but an operator's _existence_ can't be. The implementation for qutrits here, which involves changing the number of TShift operators being initialized (a concrete quantity) depending on the target state (an abstract placeholder), can't be made jittable. 

In the future, when more parametrized qutrit operations are added to PennyLane, we can hopefully add JIT compatibility similar to what has been implemented for qubit systems, where a placeholder parameter value is dependent on a placeholder target state value.

Reply:
I see, thank you so much for taking the time to explain this! Maybe I will help contribute to the future JIT comptability :)